autorest/azure/devicetoken_test.go

```diff
@@ -0,0 +1,222 @@
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Request for Device Authorization Code resulted in Status)
+	}
+}
+
+func TestDeviceCodeReturnsErrorIfCannotDeserializeDeviceCode(t *testing.T) {
+	gibberishJSON := strings.Replace(MockDeviceCodeResponse, "expires_in", "\":, :gibberish", -1)
+	sender := mocks.NewSender()
+	sender.EmitContent(gibberishJSON)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get error") // (expecting Failed to decode Device Authorization Code json)
+	}
+}
+
+func deviceCode() *DeviceCode {
+	var deviceCode DeviceCode
+	json.Unmarshal([]byte(MockDeviceCodeResponse), &deviceCode)
+	deviceCode.Resource = to.StringPtr("resource")
+	return &deviceCode
+}
+
+func TestDeviceTokenReturns(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(MockDeviceTokenResponse)
+	sender.EmitStatus("OK", 200)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err != nil {
+		t.Errorf("got error unexpectedly")
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfSendingFails(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitErrors(1)
+	sender.SetError(fmt.Errorf("this is an error"))
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while exchanging device code for token)
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfServerError(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitStatus("Internal Server Error", 500)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get error") // (expecting Exchanging Device Code for OAuth token resulted in Status)
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfCannotDeserializeDeviceToken(t *testing.T) {
+	gibberishJSON := strings.Replace(MockDeviceTokenResponse, "expires_in", ";:\"gibberish", -1)
+	sender := mocks.NewSender()
+	sender.EmitContent(gibberishJSON)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while exchanging device code for token)
+	}
+}
+
+func errorDeviceTokenResponse(message string) string {
+	return `{ "error": "` + message + `" }`
+}
+
+func TestDeviceTokenReturnsErrorIfAuthorizationPending(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(errorDeviceTokenResponse("authorization_pending"))
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	go func() {
+		time.Sleep(1 * time.Second)
```

Comment:
I'm usually against sleeps in tests. Instead perhaps you can create a new mock that gets a list of responses (`["authorization_pending", "access_denied"]`) and on the first request, returns the first response, on the 2nd returns the second and so on.